File: YYReader/Parsing/QidiySourceAdapter.swift
import Foundation
import SwiftSoup

struct QidiySourceAdapter: NovelSourceAdapter {
    func canHandle(_ document: LoadedHTML) -> Bool {
        let host = document.finalURL.host?.lowercased() ?? ""
        return host == "qidiy.com" || host.hasSuffix(".qidiy.com")
    }

    func parseChapterPage(_ loaded: LoadedHTML) throws -> ParsedChapterPage {
        let document = try HTMLParsingSupport.document(from: loaded)
        guard let content = try document.select("#content").first(),
              let heading = try document.select("h1.title").first() else {
            throw NovelParsingError.noReadableContent
        }

        let rawTitle = try heading.text().trimmingCharacters(in: .whitespacesAndNewlines)
        let title = HTMLParsingSupport.replacingRegex(
            "\\s*[（(]第\\d+/\\d+页[）)]\\s*$",
            in: rawTitle,
            with: ""
        )
        let rawParagraphs = try HTMLParsingSupport.paragraphs(from: content)
        let paragraphs = cleanChapterParagraphs(rawParagraphs, title: title)
        guard !paragraphs.isEmpty else { throw NovelParsingError.noReadableContent }

        let catalogURL = try HTMLParsingSupport.link(
            in: document,
            matching: ["章节列表", "目录"],
            baseURL: loaded.finalURL
        )
        let previous = try HTMLParsingSupport.link(
            in: document,
            matching: ["上一章"],
            baseURL: loaded.finalURL
        )
        let next = try HTMLParsingSupport.link(
            in: document,
            matching: ["下一章"],
            baseURL: loaded.finalURL
        )
        let nextPage = try HTMLParsingSupport.link(
            in: document,
            matching: ["下一页"],
            baseURL: loaded.finalURL
        )

        let bookLink = try document.select("a").array().first { anchor in
            guard let url = HTMLParsingSupport.absoluteURL(for: anchor, relativeTo: loaded.finalURL) else {
                return false
            }
            let text = (try? anchor.text()) ?? ""
            return url.path.range(of: "^/book/\\d+/?$", options: .regularExpression) != nil
                && !["上一章", "下一章", "章节列表", "开始阅读"].contains(text)
        }
        let bookTitle = try bookLink?.text().trimmingCharacters(in: .whitespacesAndNewlines)

        return ParsedChapterPage(
            title: title,
            bookTitle: bookTitle?.isEmpty == false ? bookTitle : nil,
            author: nil,
            paragraphs: paragraphs,
            catalogURL: catalogURL,
            previousChapterURL: previous == catalogURL ? nil : previous,
            nextChapterURL: next,
            nextPageURL: nextPage
        )
    }

    func parseCatalogPage(_ loaded: LoadedHTML) throws -> ParsedBookCatalog {
        let document = try HTMLParsingSupport.document(from: loaded)
        let headings = try document.select("h1").array()
        let title = try headings.first { try $0.className() != "logo" }?.text()
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard let title, !title.isEmpty else { throw NovelParsingError.missingCatalog }

        let author = try document.select("p").array()
            .compactMap { try? $0.text() }
            .first { $0.range(of: "^作者[：:]", options: .regularExpression) != nil }
            .map { HTMLParsingSupport.replacingRegex("^作者[：:]\\s*", in: $0, with: "") }
            ?? "未知作者"

        var seen = Set<String>()
        var chapters: [ChapterSeed] = []
        for anchor in try document.select("ul.section-list a").array() {
            let chapterTitle = try anchor.text().trimmingCharacters(in: .whitespacesAndNewlines)
            guard chapterTitle.range(of: "第.+[章回节]", options: .regularExpression) != nil,
                  let url = HTMLParsingSupport.absoluteURL(for: anchor, relativeTo: loaded.finalURL),
                  seen.insert(url.absoluteString).inserted else {
                continue
            }
            let index = HTMLParsingSupport.chapterNumber(in: chapterTitle) ?? (chapters.count + 1)
            chapters.append(ChapterSeed(title: chapterTitle, url: url, sortIndex: index))
        }

        let nextPage = try HTMLParsingSupport.link(
            in: document,
            matching: ["下一页"],
            baseURL: loaded.finalURL
        )

        return ParsedBookCatalog(
            title: title,
            author: author,
            chapters: chapters.sorted { lhs, rhs in
                if lhs.sortIndex == rhs.sortIndex { lhs.url.absoluteString < rhs.url.absoluteString }
                else { lhs.sortIndex < rhs.sortIndex }
            },
            nextPageURL: nextPage
        )
    }

    private func cleanChapterParagraphs(_ paragraphs: [String], title: String) -> [String] {
        let pageMarker = NSRegularExpression.escapedPattern(for: title) + "\\s*\\(第\\d+/\\d+页\\)\\s*"
        return paragraphs.compactMap { paragraph in
            var value = paragraph
            value = HTMLParsingSupport.replacingRegex("read\\d+\\(\\);?", in: value, with: "")
            value = HTMLParsingSupport.replacingRegex(pageMarker, in: value, with: "")
            value = HTMLParsingSupport.replacingRegex("[（(]本章未完[^）)]*[）)]", in: value, with: "")
            value = HTMLParsingSupport.normalize(value)
            if value.isEmpty || value.contains("加入书签，方便阅读") { return nil }
            return value
        }
    }
}

File: windows/YYReader.Windows.Core/Parsing/QidiySourceAdapter.cs
using System.Text.RegularExpressions;
using AngleSharp.Dom;
using YYReader.Windows.Core.Models;

namespace YYReader.Windows.Core.Parsing;

public sealed class QidiySourceAdapter : INovelSourceAdapter
{
    public bool CanHandle(LoadedHtml document)
    {
        var host = document.FinalUri.DnsSafeHost.ToLowerInvariant();
        return host == "qidiy.com" || host.EndsWith(".qidiy.com", StringComparison.Ordinal);
    }

    public ParsedChapterPage ParseChapterPage(LoadedHtml loaded)
    {
        var document = HtmlParsingSupport.ParseDocument(loaded);
        var content = document.QuerySelector("#content");
        var heading = document.QuerySelector("h1.title");
        if (content is null || heading is null)
        {
            throw new NovelParsingException(NovelParsingErrorKind.NoReadableContent);
        }

        var rawTitle = HtmlParsingSupport.Normalize(heading.TextContent);
        var title = Regex.Replace(rawTitle, @"\s*[（(]第\d+/\d+页[）)]\s*$", string.Empty).Trim();
        var paragraphs = CleanChapterParagraphs(HtmlParsingSupport.Paragraphs(content), title);
        if (paragraphs.Count == 0)
        {
            throw new NovelParsingException(NovelParsingErrorKind.NoReadableContent);
        }

        var catalogUrl = HtmlParsingSupport.Link(document, ["章节列表", "目录"], loaded.FinalUri);
        var previous = HtmlParsingSupport.Link(document, ["上一章"], loaded.FinalUri);
        var next = HtmlParsingSupport.Link(document, ["下一章"], loaded.FinalUri);
        var nextPage = HtmlParsingSupport.Link(document, ["下一页"], loaded.FinalUri);
        var bookLink = document.QuerySelectorAll("a")
            .Select(anchor => (anchor, url: HtmlParsingSupport.AbsoluteUrl(anchor, loaded.FinalUri)))
            .FirstOrDefault(pair => pair.url is not null
                && Regex.IsMatch(pair.url.AbsolutePath, @"^/book/\d+/?$", RegexOptions.CultureInvariant)
                && !new[] { "上一章", "下一章", "章节列表", "开始阅读" }.Contains(HtmlParsingSupport.Normalize(pair.anchor.TextContent)));
        var bookTitle = bookLink.anchor is null ? null : HtmlParsingSupport.Normalize(bookLink.anchor.TextContent);

        return new ParsedChapterPage(
            title,
            string.IsNullOrWhiteSpace(bookTitle) ? null : bookTitle,
            null,
            paragraphs,
            catalogUrl,
            previous is not null && previous == catalogUrl ? null : previous,
            next,
            nextPage);
    }

    public ParsedBookCatalog ParseCatalogPage(LoadedHtml loaded)
    {
        var document = HtmlParsingSupport.ParseDocument(loaded);
        var title = document.QuerySelectorAll("h1")
            .FirstOrDefault(heading => !heading.ClassList.Contains("logo"))
            ?.TextContent.Trim();
        if (string.IsNullOrWhiteSpace(title))
        {
            throw new NovelParsingException(NovelParsingErrorKind.MissingCatalog);
        }

        var authorText = document.QuerySelectorAll("p")
            .Select(paragraph => HtmlParsingSupport.Normalize(paragraph.TextContent))
            .FirstOrDefault(text => Regex.IsMatch(text, @"^作者[：:]", RegexOptions.CultureInvariant));
        var author = string.IsNullOrWhiteSpace(authorText)
            ? "未知作者"
            : Regex.Replace(authorText, @"^作者[：:]\s*", string.Empty).Trim();

        var sections = document.QuerySelectorAll("ul.section-list")
            .Select(section => ChapterSeeds(section, loaded.FinalUri))
            .Where(section => section.Count > 0)
            .ToArray();
        var chapters = sections.OrderByDescending(section => section.Count)
            .ThenBy(section => section.Select(seed => HtmlParsingSupport.ChapterNumber(seed.Title) ?? int.MaxValue).DefaultIfEmpty(int.MaxValue).Min())
            .FirstOrDefault() ?? new List<ChapterSeed>();

        return new ParsedBookCatalog(
            title,
            author,
            chapters,
            HtmlParsingSupport.Link(document, ["下一页"], loaded.FinalUri));
    }

    private static List<ChapterSeed> ChapterSeeds(IElement section, Uri baseUri)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        var result = new List<ChapterSeed>();
        foreach (var anchor in section.QuerySelectorAll("a"))
        {
            var title = HtmlParsingSupport.Normalize(anchor.TextContent);
            var url = HtmlParsingSupport.AbsoluteUrl(anchor, baseUri);
            if (!Regex.IsMatch(title, @"第.+[章回节]", RegexOptions.CultureInvariant)
                || url is null
                || !HtmlParsingSupport.IsSameOrigin(url, baseUri)
                || !seen.Add(url.AbsoluteUri))
            {
                continue;
            }

            result.Add(new ChapterSeed(title, UrlCanonicalizer.CanonicalizeChapter(url), result.Count + 1));
        }

        return result;
    }

    private static List<string> CleanChapterParagraphs(IEnumerable<string> paragraphs, string title)
    {
        var pageMarker = $"{Regex.Escape(title)}\\s*\\(第\\d+/\\d+页\\)\\s*";
        return paragraphs
            .Select(paragraph => Regex.Replace(paragraph, @"read\d+\(\);?", string.Empty, RegexOptions.IgnoreCase))
            .Select(paragraph => Regex.Replace(paragraph, pageMarker, string.Empty))
            .Select(paragraph => Regex.Replace(paragraph, @"[（(]本章未完[^）)]*[）)]", string.Empty))
            .Select(HtmlParsingSupport.Normalize)
            .Where(paragraph => paragraph.Length > 0 && !paragraph.Contains("加入书签，方便阅读", StringComparison.Ordinal))
            .ToList();
    }
}
